<img width="500" src="https://github.com/maxgmarin/pupmapper/raw/main/Images/pupmapper.logo.png" alt="pupmapper logo">

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
![Static Badge](https://img.shields.io/badge/language-Python_3-blue)
<!---[![Build Status]()]()
[![github release version]()]()
[![DOI]()]()
--->

Pupmapper: A **P**ile**up** **Map**pability Calculato**r** 

<!---
> TBD Reference
--->


[TOC]: #
## Table of Contents
- [Motivation](#motivation)
- [Installation](#installation)
  - [Install locally](#install-locally)
  - [`pip`](#pip)
- [Basic usage](#basic-usage)
  - [test data set](#analyzing-included-test-data-set)
- [Full usage](#full-usage)


## Motivation

Pileup mappability is a useful metric for accessing sequence uniqueness within a genome and identifying repetive regions to remove from analysis when using short-read sequencing. It is calculated as the mean k-mer mappability of all k-mers overlapping a given position. This tool is meant to make it as easy as possible to quickly calculate pileup mappability from k-mer mappability values generated by the [Genmap](https://github.com/cpockrandt/genmap) software.


## Installation
### Install locally
`pupmapper` can be installed by cloning this repository and installing with `pip`.

```
git clone git@github.com:maxgmarin/pupmapper.git

cd pgqc

pip install . 
```

### `pip`
```
pip install pupmapper
```

## Basic usage
```
pupmapper -i kmap.K50E0.bedgraph -o pupmap.K50E0.bedgraph -k 50
```

The above command will run `pupmapper` on the input k-mer mappabilities (k= 50 bp, E = 4 mismatches) and output pileup mappability scores. 



### Analyzing included test sequence

If you wish to run an `panqc nrc` on a small test sequence, you can run the following commands:

```
cd tests/data

pupmapper -i ${in_kmap} -o ${out_pupmap} -k 4
```

## Full usage
```
$ pupmapper --help

usage: pupmapper [-h] -i INPUT -o OUTPUT -k KMER_SIZE

Tool for calculating genome wide pileup mappability based on k-mer mappability values

optional arguments:
  -h, --help            show this help message and exit
  -i, --input INPUT
                        Input k-mer mappability values in bedgraph format (.bedgraph). Ideally,
                        generated with genmap software
  -o, --output OUTPUT
                        Output table of pileup mappability (.bedgraph)
  -k --kmer_size KMER_SIZE
                        k-mer size (bp) used to generate the input k-mer mappability values
....
```


